File: src/main/java/first/engine/Engine.java
package first.engine;

import java.util.concurrent.Future;
import java.util.function.Consumer;
import java.util.function.Function;

public class Engine {
    public static <IN extends Object, OUT extends Object> void crab(Function<IN, OUT> s) {

    }

    public static <CTX extends Object> void send(CTX ctx, Consumer<CTX> s) {

    }

    public static <CTX extends Object, R extends Object> Future<R> send2(CTX ctx, Consumer<CTX> s) {
        return null;
    }

    public static <CTX extends Object, R extends Object> void send3(CTX ctx, Consumer<CTX> s) {
    }

    public static <CTX extends Object> void spawn(CTX ctx, Consumer<CTX> s) {

    }

    public static <CTX extends Object, R extends Object> void receive(Consumer<CTX> s, Future<R> p) {

    }

    public static <CTX extends Object> void ok(CTX ctx) {

    }
}

File: src/main/kotlin/first/engine/Engine.kt
package first.engine

fun <CTX> send(
    ctx: CTX,
    state: (CTX) -> Unit
) {

}

fun <CTX, R> send2(
    ctx: CTX,
    state: (CTX) -> Result<R>
) {

}

fun <CTX> move(
    ctx: CTX,
    state: (CTX) -> Unit
) {

}

fun <CTX> send(
    ctx: CTX,
    state: (CTX) -> Unit,
    receiver: (Unit) -> Unit
) {

}

fun <CTX, R> receive(
    state: (CTX) -> Result<R>
): Result<R>? {
    return null;
}

fun receive2(
    state: Function<Unit>
): Result2? {
    return null;
}

sealed class Result<R>()

class Ok<CTX>(val ctx: CTX) : Result<CTX>()
class Error<ERR>(val err: ERR) : Result<ERR>()

sealed class Result2()

class Success<CTX>(val ctx: CTX) : Result2()
class Error2<ERR>(val err: ERR) : Result2()

fun <R> complete(
    ctx: R
) {

}

fun <R> complete(
    ctx: R,
    state: Function<Unit>
) {

}

annotation class Starting
annotation class Terminal